File: javascript/78_-_Subsets.js
/**
 * 78. Subsets - Medium
 * Given an integer array nums of unique elements, return all possible subsets (the power set).
 * The solution set must not contain duplicate subsets. Return the solution in any order.
 */

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function(nums) {
    const ans = [];
    backtrack(nums, 0, [], ans);
    return ans;
};

var backtrack = function(nums, start, subset, powerset) {
    powerset.push(Array.from(subset));
    
    for (let i = start; i < nums.length; ++i) {
        subset.push(nums[i]);
        backtrack(nums, i + 1, subset, powerset);
        subset.pop();
    }
};

File: csharp/78_-_Subsets.cs
/**
 * 78. Subsets - Medium
 * Given an integer array nums of unique elements, return all possible subsets (the power set).
 * The solution set must not contain duplicate subsets. Return the solution in any order.
 */

public class Solution {
    public IList<IList<int>> Subsets(int[] nums) {
        IList<IList<int>> powerset = new List<IList<int>>();
        IList<int> subset = new List<int>();
        
        Backtrack(nums, 0, subset, powerset);
        return powerset;
    }
    
    public void Backtrack(int[] nums, int start, IList<int> subset, IList<IList<int>> powerset) {
        powerset.Add(new List<int>(subset));
        
        for (int i = start; i < nums.Length; ++i) {
            subset.Add(nums[i]);
            Backtrack(nums, i + 1, subset, powerset);
            subset.RemoveAt(subset.Count - 1);
        }
    }
}